dstep/translator/HeaderIndex.d

```diff
@@ -0,0 +1,267 @@
+                visit(header);
+        }
+
+        return sorted;
+    }
+
+    private size_t[][] transitiveClosure()
+    {
+        import std.range;
+
+        auto sorted = topologicalSort();
+        auto closure = new size_t[][sorted.length];
+
+        auto includedTo = new size_t[][headers_.length];
+
+        foreach (inclusion; directInclusions.byKey)
+            includedTo[inclusion[1]] ~= inclusion[0];
+
+        foreach (header; sorted)
+        {
+            closure[header] ~= header;
+
+            foreach (transitiveHeader; closure[header])
+            {
+                foreach (includedHeader; includedTo[header])
+                    closure[includedHeader] ~= transitiveHeader;
+            }
+        }
+
+        return closure;
+    }
+}
+
+class HeaderIndex
+{
+    private IncludeGraph includeGraph_;
+    private string[string] stdLibPaths;
+    private string mainFilePath;
+
+    public this(TranslationUnit translationUnit)
+    {
+        import std.algorithm;
+
+        auto directives = translationUnit.cursor.children
+            .filter!(cursor => cursor.kind == CXCursorKind.inclusionDirective);
+
+        includeGraph_ = new IncludeGraph(directives);
+        mainFilePath = translationUnit.spelling.asAbsNormPath;
+        this(directives);
+    }
+
+    private this(T)(T directives)
+    {
+        import std.random;
+        import std.range;
+
+        stdLibPaths = resolveStdLibPaths(directives);
+    }
+
+    bool isFromStdLib(string path, string header)
+    {
+        auto headerPath = header in stdLibPaths;
+
+        if (headerPath is null)
+            return false;
+
+        return includeGraph.isReachableThrough(
+            path.asAbsNormPath,
+            *headerPath,
+            mainFilePath);
+    }
+
+    bool isFromStdLib(Cursor cursor, string header)
+    {
+        return isFromStdLib(cursor.file.name, header);
+    }
+
+    IncludeGraph includeGraph()
+    {
+        return includeGraph_;
+    }
+
+    private string[string] resolveStdLibPaths(T)(T directives) const
+    {
+        immutable uint[string] stdlib = [
+            "assert.h" : 89,
+            "ctype.h" : 89,
+            "errno.h" : 89,
+            "float.h" : 89,
+            "limits.h" : 89,
+            "locale.h" : 89,
+            "math.h" : 89,
+            "setjmp.h" : 89,
+            "signal.h" : 89,
+            "stdarg.h" : 89,
+            "stddef.h" : 89,
+            "stdio.h" : 89,
+            "stdlib.h" : 89,
+            "string.h" : 89,
+            "time.h" : 89,
+            "iso646.h" : 95,
+            "wctype.h" : 95,
+            "wchar.h" : 95,
+            "complex.h" : 99,
+            "fenv.h" : 99,
+            "inttypes.h" : 99,
+            "tgmath.h" : 99,
+            "stdint.h" : 99,
+            "stdbool.h" : 99,
+            "stdnoreturn.h" : 11,
+            "threads.h" : 11,
+            "uchar.h" : 11,
+            "stdatomic.h" : 11,
+            "stdalign.h" : 11
+        ];
```

Comment:
I assume the values are the C standard the header was added to? As far as I can see the value is not used, to you plan to use it in the future or should this be replaced with a set?

Reply:
I will probably not use it, I did it this wat because there is no literal for set... I would need to create an array and the convert to the set.

Reply:
Ah, I see. Fair enough.